### PR TITLE
StateMachineTransitionWalker and StateMachine docs

### PR DESCRIPTION
**Pull request by JoshuaBehrens**
### 1. Why is this change necessary?
After I [answered a question on SO](https://stackoverflow.com/a/63863270/1203643) I recalled a time where I made a state machine. To answer the question I tried to refer to the docs but there aren't any 😞 

As there are no docs about the state machine I wrote a paper about it and my utility class that makes it easier to transition from any state to an other by going a legal way.

### 2. What does this change do, exactly?
Add a utility class to transition from state to state for any entity (that has a single primary key 💭).
Add a big how to entry for state machine handling.

### 3. Please link to the relevant issues (if any).
https://stackoverflow.com/q/63847132/1203643

### 4. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/master/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by JoshuaBehrens** [comment]
Regarding the test results the order state was intended to be writable. Can @shopwareBot confirm this?

File: src/Core/System/StateMachine/UnwalkablePathException.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\System\StateMachine\Exception;
+
+class UnwalkablePathException extends \RuntimeException
```
Comment (mitelg) [inline src/Core/System/StateMachine/UnwalkablePathException.php]: please extend from the ShopwareHttpException

File: src/Docs/Resources/current/50-how-to/360-state-machine.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+### How to create a custom state machine
+
+The most reliable way to ship your state machine is by providing a migration that creates the state machine:
+
+```php
+use Shopware\Core\Defaults;
+use Shopware\Core\Framework\Uuid\Uuid;
+
+$stateMachineId = Uuid::randomBytes();
+$now = date_create()->format(Defaults::STORAGE_DATE_TIME_FORMAT);
```
Comment (mitelg) [inline src/Docs/Resources/current/50-how-to/360-state-machine.md]:
```suggestion
$now = (new \DateTime())->format(Defaults::STORAGE_DATE_TIME_FORMAT);
```

**Comment by JoshuaBehrens** [comment]
I rebased and fixed the tests. Hf merging :) @raknison

**Comment by philipreinken** [comment]
Hey @JoshuaBehrens , unfortunately we can't merge this PR, since the changes go against the point of a state machine somewhat. One should explicitly not be able to "jump" anywhere except directly neighbouring states. Even if you're not actually doing this (since you're automatically applying the needed actions to get there), you still risk applying a random state/action in between when there are multiple paths to a target state.

**Comment by JoshuaBehrens** [comment]
@philipgatzka I'd still like to have some more docs about it. It is ok to repost them?

If this is not intended I'd rather publish a library that handles this as I need this for data transfer. In my case: An ERP is not aware of the steps in shopware and should not care about it when it has a different state and wants to set it.

**Comment by mitelg** [comment]
yes, please repost the docs 💙

**Comment by JoshuaBehrens** [comment]
@philipgatzka What about an unclear path exception? This would still validate the path so persons have to take the choice but it is easier to use. I don't want to enforce this into shopware just want to understand what the target of this design decision is. Right now it feels complicated by design in usage

**Comment by philipreinken** [comment]
I think the reason for this decision is implementing the "Workflow dependant" design goal described [here](https://docs.shopware.com/en/shopware-platform-dev-en/references-internals/core/checkout-process/order#design-goals):
> Orders states change in a defined, predictable and configurable way - other state transitions are blocked.

**Comment by JoshuaBehrens** [comment]
I see. So an unclear path exception would be a solution to follow the rules but simplify the process, right?

**Pull request closed without merging**